database.cc

```diff
@@ -2321,6 +2321,71 @@ future<> database::drain() {
     co_await _commitlog->shutdown();
 }
 
+future<std::unordered_map<sstring, std::unordered_map<std::pair<sstring, sstring>, db::snapshot_ctl::snapshot_details, utils::tuple_hash>>> database::get_snapshot_details() {
+    // We require a seastar thread to run `lister::scan_dir`s
+    co_return co_await seastar::async([this] () {
+        std::vector<sstring> data_dirs = _cfg.data_file_directories();
+        // auto dirs_only_entries_ptr =
+        //     make_lw_shared<lister::dir_entry_types>(lister::dir_entry_types{directory_entry_type::directory});
+        auto dirs_only_entries = lister::dir_entry_types{directory_entry_type::directory};
+        std::unordered_map<sstring, std::unordered_map<std::pair<sstring, sstring>, snapshot_ctl::snapshot_details, utils::tuple_hash>> details;
+
+        for (auto& datadir : data_dirs) {
+            lister::scan_dir(datadir, dirs_only_entries, [this, &dirs_only_entries, &details] (fs::path parent_dir, directory_entry de) mutable {
+                // KS directory
+                sstring ks_name = de.name;
+
+                return lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &dirs_only_entries, &details, ks_name = std::move(ks_name)] (fs::path parent_dir, directory_entry de) mutable {
+                    // CF directory
+                    sstring cf_name = de.name.substr(0, de.name.find("-"));
+                    fs::path cf_dir = parent_dir / de.name;
+
+                    return lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &dirs_only_entries, &details, ks_name, cf_name = std::move(cf_name), cf_dir = std::move(cf_dir)] (fs::path parent_dir, directory_entry de) mutable {
+                        // "snapshots" directory
+
+                        return lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &details, ks_name, cf_name, cf_dir] (fs::path parent_dir, directory_entry de) {
```

Comment:
Why suddenty not performing `move`s on `ks_name` etc. in the lambda capture?

Reply:
BTW. The logic inside this `scan_dir` (which handles a single snapshot) could really benefit (in terms of readability) from being extracted to a separate function.

Reply:
> Why suddenty not performing moves on ks_name etc. in the lambda capture?

Nested `scan_dirs` are like nested loops, so we don't want to move a value from an outer loop into an inner loop, cause we will still need it in the next iteration of the outer one. However, I might be able to pass by reference there and avoid excessive copies anyway.